### PR TITLE
expand webpki verifier CRL support

### PR DESCRIPTION
**Pull request by cpu**
This branch achieves three top-level goals:

1. It expands the existing webpki client certificate verifier builder to allow configuring the revocation checking depth (by default, full chain when CRLs are provided, optionally only end-entity), and to control how unknown revocation status is handled (by default, treating unknown revocation status as an error when CRLs are provided, optionally allowing unknown).
2. It reworks the webpki server certificate verifier to be constructed using a builder, matching the client verifier.
3. It adds optional CRL support to the server certificate verifier, allowing users to opt in to providing CRLs and controlling how revocation will be handled (depth, unknown status, etc). This support will be helpful towards resolving #1541.

### CODE REVIEW

File: rustls/tests/common/mod.rs [hunk 1]
```diff
@@ -284,8 +302,8 @@ pub fn make_server_config_with_kx_groups(
 
 pub fn get_client_root_store(kt: KeyType) -> Arc<RootCertStore> {
     let mut roots = kt.get_chain();
-    // drop server cert
-    roots.drain(0..1);
+    // drop server cert and intermediate
+    roots.drain(0..2);
```
Comment (cpu) [inline rustls/tests/common/mod.rs:306]: This change is worth a bit of discussion. Previously we were constructing the client root store by reading the `client.fullchain` file (containing EE cert, intermediate and root) and then discarding just the EE cert. The net effect is that we build a root store here that contains _both_ the intermediate and the root.

We use this root store when constructing a client certificate verifier for two purposes:

1. For populating a list of subjects we send in the [certificate_authorities](https://www.rfc-editor.org/rfc/rfc8446#section-4.2.4) extension of a `CertificateRequest` message (TLS 1.2 is handled similarly).
2. For populating a set of trust anchors we provide to webpki when verifying the client certificate.

For the first purpose it _might_ make sense to include the intermediate subjects. I'm not 100% sure. Reading the RFC text I think it could go either way.

For the second purpose it doesn't make sense to include the intermediate cert. We end up using it as a trust anchor which isn't appropriate: it's an intermediate certificate itself issued by a trust anchor. If we try path-building with the EE cert, the intermediates, and these trust anchors we never look at the intemediates: our built path is EE->TA (where the TA is the first intermediate). This means checking revocation status of the intermediate is never performed: we consider it a TA and those aren't covered by CRLs.

In this branch to make it feasible to test the intermediate revocation checking I've changed this function to only return the root TA, not the intermediate. As a result one other unit test (`client_cert_resolve`) that expected to see the intermediate subject in the `certificate_authorities` ext has to be updated to only expect the root TA subject.

File: rustls/src/error.rs [hunk 2]
```diff
@@ -372,7 +375,7 @@ impl From<CertificateError> for AlertDescription {
             // certificate_unknown
             //  Some other (unspecified) issue arose in processing the
             //  certificate, rendering it unacceptable.
-            Other(_) => Self::CertificateUnknown,
+            Other(_) | UnknownRevocationStatus => Self::CertificateUnknown,
```
Comment (cpu) [inline rustls/src/error.rs:378]: I'm not sure this mapping from the new unknown revocation status certificate error to alert makes sense. Suggestions welcome.

File: rustls/src/webpki/verify.rs [hunk 3]
```diff
@@ -34,17 +39,14 @@ pub fn verify_server_cert_signed_by_trust_anchor(
...
-            supported_algs,
-            &roots.roots,
-            intermediates,
-            now,
-            webpki::KeyUsage::server_auth(),
-            None, // no CRLs
-        )
-        .map_err(pki_error)
-        .map(|_| ())
+    verify_server_cert_signed_by_trust_anchor_impl(
```
Comment (cpu) [inline rustls/src/webpki/verify.rs:42]: We want to be able to provide the `revocation` argument to Webpki's `verify_for_usage` fn when we're crate-internal, but can't have the webpki revocation types bleed into the public API.

This approach, having the public helper fn not support revocation checking, and deferring to an internal impl that does was the best I could come up with :-(

As discussed in https://github.com/rustls/webpki/issues/159#issuecomment-1739695300 moving the revocation types into `pki-types` isn't very appealing, but I'm open to other suggestions.

**Comment by cpu** [comment]
> codecov/patch — 95.15% of diff hit (target 96.46%)

I'll see if I can address some of these coverage gaps (likely ~Mon).

File: rustls/src/webpki/verify.rs [hunk 4]
```diff
@@ -437,6 +437,18 @@ impl fmt::Debug for WebPkiSupportedAlgorithms {
     }
 }
 
+/// wrapper around internal representation of a parsed certificate. This is used in order to avoid parsing twice when specifying custom verification
```
Comment (ctz) [inline rustls/src/webpki/verify.rs:499]: not changed by this commit, but should start with a capital letter, start with a "headline" sentence, and be manually wrapped

```suggestion
/// Wrapper around internal representation of a parsed certificate.
///
/// This is used in order to avoid parsing twice when specifying custom verification.
```

File: rustls/src/webpki/client_verifier_builder.rs [hunk 5]
```diff
@@ -195,13 +195,13 @@ mod tests {
     }
 
     #[test]
-    fn test_noauth() {
+    fn test_client_verifier_noauth() {
```
Comment (djc) [inline rustls/src/webpki/client_verifier_builder.rs]: Pre-existing issue, but maybe split `noauth` into `no_auth` while you're in here?

File: rustls/src/webpki/client_verifier_builder.rs [hunk 6]
```diff
@@ -43,6 +50,20 @@ impl ClientCertVerifierBuilder {
         self
     }
 
+    /// When CRLs are provided with [`with_crls`][Self::with_crls] only check the end entity
```
Comment (djc) [inline rustls/src/webpki/client_verifier_builder.rs]: Maybe tighten up the docstring headline a bit? (I'd start by removing ", not the entire chain", which feels like something that doesn't need to be in the headline.)

File: rustls/src/error.rs (hunk 2 above)
Comment (djc) [reply rustls/src/error.rs:378]: Just in terms of names, `BadCertificate` seems more appropriate than `CertificateUnknown`? Did you check what BoringSSL/OpenSSL do?

File: rustls/src/webpki/client_verifier_builder.rs [hunk 7]
```diff
@@ -73,6 +80,19 @@ impl ClientCertVerifierBuilder {
         self
     }
 
+    /// When CRLs are provided with [`with_crls`][Self::with_crls] if it isn't possible to
```
Comment (djc) [inline rustls/src/webpki/client_verifier_builder.rs]: Nit: tighten up the headline?

File: rustls/src/webpki/verify.rs [hunk 8]
```diff
@@ -278,13 +279,16 @@ impl WebPkiClientVerifier {
     /// client certificate validation.
     /// `revocation_check_depth` controls which certificates have their revocation status checked
     /// when `crls` are provided.
+    /// `unknown_revocation_policy` controls how certificates with an unknown revocation status
```
Comment (djc) [inline rustls/src/webpki/verify.rs:333]: How does this look in rustdoc? Should we use a bullet list here?

File: rustls/src/webpki/mod.rs [hunk 9]
```diff
@@ -3,12 +3,12 @@ use alloc::sync::Arc;
 use crate::error::{CertRevocationListError, CertificateError, Error};
 
 mod anchors;
-mod client_verifier_builder;
+mod verifier_builders;
```
Comment (djc) [inline rustls/src/webpki/mod.rs:6]: Alternative proposal: go the other way and split `client_verifier_builder` and `verify` into `client_verifier` and `server_verifier` builder modules that contain the builders? In general I don't love it when our builders live in a different file from the type they build.

File: rustls/src/webpki/verifier_builders.rs [hunk 10]
```diff
@@ -159,7 +159,7 @@ impl ClientCertVerifierBuilder {
 /// instead of constructing a builder.
 #[derive(Debug, Clone)]
 #[non_exhaustive]
-pub enum ClientCertVerifierBuilderError {
+pub enum CertVerifierBuilderError {
```
Comment (djc) [inline rustls/src/webpki/verifier_builders.rs:283]: Naming nit: how about tightening this up to `VerifierBuilderError`? I feel like the `Cert` prefix is fairly superfluous in this context?

File: rustls/src/webpki/verify.rs [hunk 11]
```diff
@@ -275,7 +275,7 @@ impl WebPkiClientVerifier {
     /// Construct a new `WebpkiClientVerifier`.
     ///
     /// `roots` is the list of trust anchors to use for certificate validation.
-    /// `crls` are an iterable of owned certificate revocation lists (CRLs) to use for
+    /// `crls` are a vec of owned certificate revocation lists (CRLs) to use for
```
Comment (djc) [inline rustls/src/webpki/verify.rs:329]: I would capitalize `Vec`. (I would also write `is` rather than `are`, and maybe this deserves a bullet list, too?)

File: provider-example/src/lib.rs [hunk 12]
```diff
@@ -61,8 +61,8 @@ pub static TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256: rustls::SupportedCipherS
 pub fn certificate_verifier(
     roots: rustls::RootCertStore,
 ) -> Arc<dyn rustls::client::danger::ServerCertVerifier> {
-    Arc::new(rustls::client::WebPkiServerVerifier::new_with_algorithms(
-        roots,
-        verify::ALGORITHMS,
-    ))
+    rustls::client::WebPkiServerVerifier::builder(roots.into())
```
Comment (djc) [inline provider-example/src/lib.rs:64]: I think the commit message got truncated for this cmmit ("builder for server cert verifier, CRL support")

File: rustls/src/webpki/verify.rs (hunk 3 above)
Comment (djc) [reply rustls/src/webpki/verify.rs:42]: Seems like an okay trade-off for now.

File: rustls/src/webpki/verifier_builders.rs [hunk 13]
```diff
@@ -154,6 +156,125 @@ impl ClientCertVerifierBuilder {
...
+        Self {
+            roots,
+            crls: Vec::new(),
+            revocation_check_depth: RevocationCheckDepth::Chain,
+            unknown_revocation_policy: UnknownStatusPolicy::Deny,
+            supported_algs: None,
+        }
+    }
+
+    /// Verify the revocation state of presented client certificates against the provided
```
Comment (djc) [inline rustls/src/webpki/verifier_builders.rs:182]: Nit: tighter headlines?

File: rustls/src/webpki/verify.rs [hunk 14]
```diff
@@ -457,6 +508,40 @@ impl<'a> TryFrom<&'a CertificateDer<'a>> for ParsedCertificate<'a> {
...
+    cert.0
+        .verify_for_usage(
+            supported_algs,
+            &roots.roots,
+            intermediates,
+            now,
+            webpki::KeyUsage::server_auth(),
+            revocation,
+        )
+        .map_err(pki_error)
```
Comment (djc) [inline rustls/src/webpki/verify.rs:541]: I tend to dislike chaining combinators that operate on different variants -- maybe try binding `result` and using an explicit `match` for that, see what you think?

File: rustls/src/webpki/verify.rs [hunk 15]
```diff
@@ -151,10 +181,31 @@ impl ServerCertVerifier for WebPkiServerVerifier {
...
         let cert = ParsedCertificate::try_from(end_entity)?;
 
-        verify_server_cert_signed_by_trust_anchor(
+        let crls = borrow_crls(&self.crls);
+        let revocation = if crls.is_empty() {
+            None
+        } else {
+            // Note: unwrap here is safe because RevocationOptionsBuilder only errors when given
+            //       empty CRLs.
+            let mut builder = webpki::RevocationOptionsBuilder::new(&crls)
```
Comment (djc) [inline rustls/src/webpki/verify.rs:190]: The `borrow_crls(&self.crls)` -> `RevocationOptionsBuilder::new(&crls).unwrap()` dance looks pretty convoluted... is this what we ended up on for the client? Maybe webpki should expose some API to make this more streamlined?

**Review by djc — COMMENTED** [review]
Haven't looked at the testing commits much yet, here's a first round of feedback.

File: rustls/src/error.rs (hunk 2 above)
Comment (cpu) [reply rustls/src/error.rs]: >  Did you check what BoringSSL/OpenSSL do?

I find the BoringSSL/OpenSSL code pretty hard to follow but I'll put some legwork in and see what I can figure out. 

> Just in terms of names, BadCertificate seems more appropriate than CertificateUnknown?

Switched to `BadCertificate` in the meantime.

File: rustls/src/webpki/verify.rs (hunk 4 above)
Comment (cpu) [reply rustls/src/webpki/verify.rs]: Fixed up.

File: rustls/src/webpki/client_verifier_builder.rs (hunk 5 above)
Comment (cpu) [reply rustls/src/webpki/client_verifier_builder.rs]: Done.

File: rustls/src/webpki/client_verifier_builder.rs (hunk 6 above)
Comment (cpu) [reply rustls/src/webpki/client_verifier_builder.rs]: Done. Hopefully clearer now :crossed_fingers:

File: rustls/src/webpki/client_verifier_builder.rs (hunk 7 above)
Comment (cpu) [reply rustls/src/webpki/client_verifier_builder.rs]: Done.

File: rustls/src/webpki/verify.rs (hunk 8 above)
Comment (cpu) [reply rustls/src/webpki/verify.rs]: Switched this and the server equivalent to use bulleted lists. They're both `pub(crate)` but it doesn't hurt to make the Rustdoc more readable for `--document-private-items`.

File: rustls/src/webpki/mod.rs (hunk 9 above)
Comment (cpu) [reply rustls/src/webpki/mod.rs]: Sure, that seems sensible. I've reworked accordingly.

File: rustls/src/webpki/verifier_builders.rs (hunk 10 above)
Comment (cpu) [reply rustls/src/webpki/verifier_builders.rs]: SGTM. Done.

File: rustls/src/webpki/verify.rs (hunk 11 above)
Comment (cpu) [reply rustls/src/webpki/verify.rs]: Fixed up.

File: provider-example/src/lib.rs (hunk 12 above)
Comment (cpu) [reply provider-example/src/lib.rs:64]: Ooops. Thanks for noticing. Fixed.

File: rustls/src/webpki/verifier_builders.rs (hunk 13 above)
Comment (cpu) [reply rustls/src/webpki/verifier_builders.rs]: Done.

File: rustls/src/webpki/verify.rs (hunk 14 above)
Comment (cpu) [reply rustls/src/webpki/verify.rs]: That's fair. I think the explicit `match` reads better :+1: Updated.

File: rustls/tests/common/mod.rs (hunk 1 above)
Comment (cpu) [reply rustls/tests/common/mod.rs]: > SGTM. I'll do some tidying.

Done.

>  I could take a shot at breaking this up in a follow-up PR. I think it's distinct enough from the work here that it'd be clearer to handle separately.

Filed a separate issue for this so I can resolve this feedback thread: https://github.com/rustls/rustls/issues/1549

File: rustls/src/error.rs (hunk 2 above)
Comment (cpu) [reply rustls/src/error.rs]: > I'll put some legwork in and see what I can figure out

I believe for both OpenSSL and BoringSSL in the absence of callbacks doing something differently, the failure to find a relevant CRL for a certificate in-scope for revocation checking will be turned into a `X509_V_ERR_UNABLE_TO_GET_CRL` result ([OpenSSL](https://github.com/openssl/openssl/blob/fac61ea4618c83826b51aebf03cbc2bc3ac7b8c8/crypto/x509/x509_vfy.c#L1015-L1024), [BoringSSL](https://github.com/google/boringssl/blob/bfa8369795b7533a222a72b7a1bc928941cd66bf/crypto/x509/x509_vfy.c#L832-L843)). In both cases that error is mapped to an Unknown CA alert ([OpenSSL](https://github.com/openssl/openssl/blob/fac61ea4618c83826b51aebf03cbc2bc3ac7b8c8/ssl/statem/statem_lib.c#L1763), [BoringSSL](https://github.com/google/boringssl/blob/bfa8369795b7533a222a72b7a1bc928941cd66bf/ssl/ssl_x509.cc#L1330C1-L1330C1)).

We could do the same, but at first pass it feels to me like `BadCertificate` might be a better fit? I haven't been able to find any standards text that would suggest which is more appropriate.

Comment (cpu) [reply rustls/src/error.rs]: > We could do the same, but at first pass it feels to me like BadCertificate might be a better fit? I haven't been able to find any standards text that would suggest which is more appropriate.

Since the existing `From<CertificateError> for AlertDescription ` impl is commented as "heavily referenced" from the OpenSSL/BoringSSL implementations I decided to tack on a commit that updates our handling of `Error::UnknownRevocationStatus` to use `AlertDescription::UnknownCA` for this condition too: 155fbd6b0f4c51eb0f7b249c0ea90dba5d1d3ad4

It's probably the least surprising thing to do and I can't find strong justification for doing otherwise :shrug: We can always change it again if someone disagrees with my logic here.

File: rustls/src/webpki/verify.rs (hunk 15 above)
Comment (cpu) [reply rustls/src/webpki/verify.rs]: I have a couple branches under way that should make this more ergonomic. Since it will require some song/dance to get webpki fixed up and then a new alpha used here maybe we could look at that separately from this PR? WDYT?

Comment (cpu) [reply rustls/src/webpki/verify.rs]: > I have a couple branches under way that should make this more ergonomic.

Webpki: https://github.com/rustls/webpki/pull/203
Rustls: https://github.com/rustls/rustls/pull/1552

File: rustls/tests/api.rs [hunk 16]
```diff
@@ -1183,6 +1185,42 @@ fn client_check_server_certificate_intermediate_revoked() {
...
+#[test]
+fn client_check_server_certificate_helper_api() {
+    for kt in ALL_KEY_TYPES.iter() {
+        let chain = kt.get_chain();
+        let correct_roots = get_client_root_store(*kt);
+        let incorrect_roots = get_client_root_store(match kt {
+            KeyType::Rsa => KeyType::Ecdsa,
+            _ => KeyType::Rsa,
+        });
+        // Using the correct turst anchors, we should verify without error.
```
Comment (ctz) [inline rustls/tests/api.rs:1200]:
```suggestion
        // Using the correct trust anchors, we should verify without error.
```

**Review by ctz — APPROVED** [review]

File: rustls/tests/api.rs (hunk 16 above)
Comment (cpu) [reply rustls/tests/api.rs]: Oops. Fixed up :-)

**Comment by cpu** [comment]
> icount bench / Run icount benchmarks (pull_request) Failing after 2m

The diffs here are either very small, or speedups, so I think we're OK here.
